src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+    const std::shared_ptr<AbstractLQPNode>& lqp,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping) {
+  // We are only interested in predicate, projection, validate, aggregate and sort nodes. These only ever have one
+  // input, thus we can scan the path of nodes linearly. This makes it easy to track which projection nodes actually
+  // need to be removed, and which predicates need to be considered when patching aggregates.
+  PredicatePullUpInfo info{};
+  size_t num_projections_to_remove = 0;
+  size_t num_aggregates_to_patch = 0;
+
+  auto node = lqp;
+  while (node != nullptr) {
+    if (node->type == LQPNodeType::Projection) {
+      info.projection_nodes.emplace_back(std::static_pointer_cast<ProjectionNode>(node));
+    } else if (node->type == LQPNodeType::Predicate) {
+      const auto& predicate_node = std::static_pointer_cast<PredicateNode>(node);
+      auto maybe_predicate_info =
+          should_become_join_predicate(predicate_node, parameter_mapping, !info.aggregate_nodes.empty());
+      if (maybe_predicate_info) {
+        info.predicates.emplace_back(*maybe_predicate_info);
+        info.predicate_nodes.emplace(predicate_node);
+        // All projections/aggregates found so far need to be removed/patched
+        num_projections_to_remove = info.projection_nodes.size();
+        num_aggregates_to_patch = info.aggregate_nodes.size();
+      }
+    } else if (node->type == LQPNodeType::Aggregate) {
+      info.aggregate_nodes.emplace_back(std::static_pointer_cast<AggregateNode>(node), info.predicates.size());
+    } else if (node->type == LQPNodeType::Alias) {
+      info.alias_nodes.emplace_back(std::static_pointer_cast<AliasNode>(node));
+    } else if (node->type != LQPNodeType::Validate && node->type != LQPNodeType::Sort) {
+      // It is not safe to pull up predicates past this node, stop scanning
+      break;
+    }
+
+    DebugAssert(!node->right_input(), "Scan only implemented for nodes with one input");
+    node = node->left_input();
+  }
+
+  // Remove projections/aggregates found below the last predicate which we don't need to remove/patch.
+  info.projection_nodes.resize(num_projections_to_remove);
+  info.aggregate_nodes.resize(num_aggregates_to_patch);
+
+  return info;
+}
+
+/**
+ * Check whether an LQP node uses a correlated parameter.
+ *
+ * If the node uses a subquery, its nodes are also all checked.
+ */
+template <class ParameterPredicate>
+bool uses_correlated_parameters(const std::shared_ptr<AbstractLQPNode>& node,
+                                ParameterPredicate&& is_correlated_parameter) {
+  bool is_correlated = false;
+  for (const auto& expression : node->node_expressions) {
+    visit_expression(expression, [&](const auto& sub_expression) {
+      // We already know that the node is correlated, so we can skip the rest of the expression
+      if (is_correlated) {
+        return ExpressionVisitation::DoNotVisitArguments;
```

Comment:
When would this be called?

Reply:
This was so that the rest of the visitation was cancelled as fast as possible after we found a correlated parameter (since returning `DoNotVisitArguments` only applies to the arguments of the correlated parameter expression)